### DISCUSSION TITLE
use a websocket reverse proxy to relay OCPP messages to multiple backends

### DISCUSSION THREAD

**Discussion by scruysberghs** [lbbrhzn/ocpp#1249] [Q&A]
As stated in the title. I have a charging station that can only talk to one OCPP backend. I need it to be able to connect to a cloud based OCPP backend for invoicing etc but also want to control it locally trough the OCPP integration in Home Assistant. Has anybody looked at putting a reverse proxy (nginx of Caddy or...) in between that you can set as endpoint in the charging station and relays messages both ways? Has this bee done before?

**Comment by rsyrnicki** [comment]
I don't think that's possible. You would to maintain a connection to two backends at the same time. AFAIK websockets doesn't support that.

**Reply by scruysberghs** [reply]
I've started working on a small HA addon to do this. It will take a websocket connection from the charging station (CS) and relay messages to multiple backends (configurable from the UI) and also relay reply's from the backends to the CS.  It uses MQTT in the middle to handle message mapping and allow for multiple charging stations talking to multiple backends in the future.
Still a work in progress but let me know if you want to help test/debug.

**Reply by rsyrnicki** [reply]
If you have some success then I'd like to test it.

**Reply by scruysberghs** [reply]
I stoped working on the extra in between service. Makes more sense to add it to this existing ha ocpp integration. Adding extra config options to optionaly  relay messages from and to another central system (local or in the cloud). So 3 features we are working on:
- forwarding messages to secondary central system
- adding support for charging stations with multiple connectors
- adding support to add multiple charging stations to one ha ocpp central system

**Reply by brecht1234** [reply]
Are you still working on this? Can you give a pointer to the existing HA OCPP integration where this is being discussed/implemented?

**Reply by scruysberghs** [reply]
we have a OCPP fork in our github org that implemented these features but meanwhile have been focussing on controlling smart charging trough other (anything but OCPP) protocols (local API or MQTT or Modbus or even cloud API) so the fork hasn't been updated since last summer. We decided that as an EMS we don't want to be involved (man in the middle) in the communication with a CPO that handels financial transactions..

**Reply by jens-markussen** [reply]
@scruysberghs , I would be interested in seeing what was done (even if not complete) to introduce relaying to a backend into the HA OCPP integration. Is the fork public?

I am looking to support a similar scenario where the charger needs to stay logically connected to a CSMS backend for capturing session, doing authentication, etc.  while meanwhile I would like to use HA to monitor charging as well as control charging current (in order to align dynamically with e.g. solar cell energy production).
